decode.go

```diff
@@ -625,7 +635,17 @@ func (d *decoder) mappingStruct(n *node, out reflect.Value) (good bool) {
 		if !d.unmarshal(ni, name) {
 			continue
 		}
-		if info, ok := sinfo.FieldsMap[name.String()]; ok {
+		ns := name.String()
+		if d.strict {
+			for _, df := range doneFields {
```

Comment:
This linear search seems a little more heavyweight than it needs to be. If we keep a unique field index in each fieldInfo (the index into fieldsList could be used when building up the struct info in getStructInfo, perhaps), then doneFields could be a []bool, and checking whether a field has already been done would just require indexing into the slice.

Reply:
I nearly followed this, but got a bit lost trying to understand the `InlineMap` stuff...

Instead I changed `doneFields` to a `map[string]bool` - which I think at minimum makes the code easier to read.